### ISSUE TITLE
HLint advises to remove the TypeApplication pragma, even if it is still necessary

### ISSUE THREAD

**Issue by KommuSoft** [ndmitchell/hlint#1270]
We can write a small file that uses the `TypeApplications` language extension, such file could look like:

```
{-# LANGUAGE TypeApplications #-}

main = putStrLn ((show @ Int) 5)
```
If we work with `hlint` v3.3.1, we get as hint:
```
file.hs:1:1-33: Warning: Unused LANGUAGE pragma
Found:
  {-# LANGUAGE TypeApplications #-}
Perhaps you should remove it.
```
but removing the language pragma means that now the compiler can no longer compile the code.

This is only the case if there is a *space* between the `@` sign and the type, so `@Int` does not suggest to remove the language extension.

Likely it sees `@` as an operator with `show` and `Int` as subexpressions, and thus `Int` is then interpreted as a data constructor and `show` as an identifier.

**Comment by shayne-fletcher** [comment]
@KommuSoft  You have worked this out. If you meant type application, the `@` needs to be in "suffix" position:
```
--  ----------+---------------+------------------------------------------
--    @       |  prefix       | ITtypeApp
--            |               |   type application, e.g.  fmap @Maybe
--            |  tight infix  | ITat
--            |               |   as-pattern, e.g.  f p@(a,b) = rhs
--            |  suffix       | parse error
--            |               |   e.g. f p@ x = rhs
--            |  loose infix  | ITvarsym "@"
--            |               |   ordinary operator or type operator,
--            |               |   e.g.  f @ g, (f @)
--  ----------+---------------+------------------------------------------
```
As you've written it in your example program the `@` is in a loose infix position. In that case, despite that program exhibiting illegal term-level use of the type constructor `Int`, it is a valid parse and so hlint reports on it and on that parse tree, indeed, `TypeApplications` is an unused extension.

**Comment by KommuSoft** [comment]
@shayne-fletcher : but the strange thing is that the `ghc` compiler sees this as type application: it will not work if the `TypeApplication` extension is not enabled, and if the `TYpeApplication` *is* enabled, it works with this as a type application.

Perhaps the error is not at the side of `hlint`, but more at the side of `ghc` that is not really following the "dialect" correctly.

**Comment by googleson78** [comment]
This(allowing a space after the `@`) is no longer the case in newer ghcs, so that's probably why hlint is reporting an error:

```haskell
> show @ Int

<interactive>:2:6: error:
    Variable not in scope: (@) :: (a0 -> String) -> t0 -> t1

<interactive>:2:8: error:
    • Illegal term-level use of the type constructor ‘Int’
        imported from ‘Prelude’ (and originally defined in ‘GHC.Types’)
```

**Comment by KommuSoft** [comment]
@googleson78 : you are right: I did this again with my personal machine (and ghc-9.0.1), and then it is parsed as an infix operator.
